pkg/virt-handler/node-labeller/kvm-caps-info-plugin.go

```diff
@@ -0,0 +1,263 @@
+		Extension: uintptr(C.CapHypervSynic2),
+		MSR:       HV_X64_MSR_SCONTROL,
+		Name:      "synic2",
+	},
+	{
+		MSR:  HV_X64_MSR_TSC_FREQUENCY,
+		Name: "frequencies",
+	},
+	{
+		MSR:  HV_X64_MSR_RESET,
+		Name: "reset",
+	},
+	{
+		MSR:  HV_X64_MSR_VP_RUNTIME,
+		Name: "runtime",
+	},
+	{
+		MSR:  HV_X64_MSR_STIMER0_CONFIG,
+		Name: "synictimer",
+	},
+	{
+		MSR:  HV_X64_MSR_REENLIGHTENMENT_CONTROL,
+		Name: "reenlightenment",
+	},
+}
+
+func (s *CapScan) availableMsrs() ([]uint32, error) {
+	buffer := make([]byte, 8, 8)
+	msrsListPtr := unsafe.Pointer(&buffer[0])
+	_, err := s.kvmIOCtl(
+		uintptr(C.IoctlGetMsrIndexList),
+		uintptr(msrsListPtr))
+	if err != 0 && err != syscall.E2BIG {
+		return nil, err
+	}
+
+	length := C.msr_list_length(msrsListPtr)
+
+	bufferSize := 4 + 4*length
+
+	newBuffer := make([]byte, bufferSize, bufferSize)
+
+	newBuffer[0] = buffer[0]
+
+	msrsListPtr = unsafe.Pointer(&newBuffer[0])
+
+	_, err = s.kvmIOCtl(
+		uintptr(C.IoctlGetMsrIndexList),
+		uintptr(msrsListPtr))
+	if err != 0 && err != syscall.E2BIG {
+		return nil, err
+	}
+	msrs := make([]uint32, 0, 0)
+
+	for i := 0; i < int(length); i++ {
+		res := C.msr_list_get(
+			msrsListPtr,
+			C.int(i))
+
+		// Add this MSR.
+		msrs = append(msrs, uint32(res))
+	}
+	return msrs, nil
+}
+
+func (s *CapScan) hasCapExtension(fd uintptr, extension uintptr) bool {
+	res, _ := s.kvmIOCtl(uintptr(C.IoctlCheckExtension), extension)
+
+	// Returns: 0 if unsupported; 1 (or some other positive integer) if supported
+	// https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt
+	if res > 0 {
+		return true
+	}
+
+	return false
+}
+
+func (s *CapScan) kvmIOCtl(flag, arg uintptr) (res uintptr, err syscall.Errno) {
+	res, _, err = syscall.Syscall(syscall.SYS_IOCTL,
+		s.fd,
+		flag,
+		arg)
+	return
+}
+
+func (s *CapScan) populateCaps() error {
+	msrs, err := s.getKVMMSRs()
+	if err != nil {
+		return err
+	}
+
+	s.supportedMSRS = make(map[uint32]bool)
+	for _, kvmMsr := range msrs {
+		s.supportedMSRS[kvmMsr] = true
+	}
+	return nil
+}
+
+func (s *CapScan) getKVMMSRs() ([]uint32, error) {
+	msrs, err := s.availableMsrs()
+	if err != nil {
+		return nil, err
+	}
+	return msrs, nil
+}
+
+func (s *CapScan) exposeCapabilities() []string {
+	exposedCaps := []string{}
+	for _, capb := range CapsDesc {
+		shouldExpose := false
+		if capb.MSR != 0 {
+			_, isEnabled := s.supportedMSRS[uint32(capb.MSR)]
+			shouldExpose = isEnabled
+		}
+		if capb.Extension != 0 {
+			res := s.hasCapExtension(s.fd, capb.Extension)
+			shouldExpose = res
+		}
+
+		if shouldExpose {
+			exposedCaps = append(exposedCaps, capb.Name)
+		}
+	}
+	return exposedCaps
+
+}
+
+func (s *CapScan) getLabels() map[string]bool {
+	labels := make(map[string]bool)
+
+	devkvm, err := os.OpenFile("/dev/kvm", syscall.O_RDWR|syscall.O_CLOEXEC, 0)
+	if err != nil {
+		log.DefaultLogger().Errorf("something happend during opening kvm file: " + err.Error())
+		return labels
+	}
+	defer devkvm.Close()
+
+	s.fd = devkvm.Fd()
```

Comment:
It would be clearer to pass the file descriptor directly to function `populateCaps()` as a parameter.